### PR TITLE
fix metrics collection in mapreduce

### PR DESCRIPTION
**Pull request by abaranau**
https://issues.cask.co/browse/CDAP-1238

Couple notes:
* all mapreduce metrics emitted as gauges, as this is how mr counters work that we use
* for that to work properly, we need to emit from tasks with task id in the context, so that for aggregate values all are summed (as expected)
* metrics collector's context in tasks now has only task type and id - since metrics reported by mr program runner, all other tags present in its context (like app id, etc.)
* added emitting custom metrics in mr tests which cought couple isuues by itself, no asserting results well - without refactoring that is coming later it is really hard

### CODE REVIEW

**Comment by abaranau** [comment]
https://builds.cask.co/browse/CDAP-RBT60-1

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java [hunk 1]
```diff
@@ -245,34 +233,35 @@ public void setReducerResources(Resources resources) {
...
 
   private static MetricsCollector getMetricCollector(MetricsCollectionService service, Program program,
-                                                     MapReduceMetrics.TaskType type, String runId) {
+                                                     MapReduceMetrics.TaskType type, String runId, String taskId) {
     if (service == null) {
       return null;
     }
-    Map<String, String> tags = Maps.newHashMap(getMetricsContext(program, runId));
+    Map<String, String> tags = Maps.newHashMap();
+    // NOTE: Currently we report metrics thru mapreduce counters and to it in mapreduce program runner. It "knows" all
```
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java]: and to it in. typo?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/MapReduceCounterCollectionService.java [hunk 2]
```diff
@@ -64,6 +69,23 @@ protected void publish(Iterator<MetricValue> metrics) throws Exception {
     }
   }
 
+  public static Map<String, String> parseTags(String counterGroupName) {
+    // see publish method for format info
+
+    if (counterGroupName.startsWith("cdap.")) {
```
Comment (gokulavasan) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/MapReduceCounterCollectionService.java]: I don't understand this part. If the counter group name starts with cdap, we return an empty map?

**Comment by gokulavasan** [comment]
One small comment. And one question for more explanation. Otherwise LGTM :+1:

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/MapReduceCounterCollectionService.java (hunk 2 above)
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/MapReduceCounterCollectionService.java]: it's more like an assertion. If it doesn't start from "cdap." then we cannot get any information about the context. I think I can even turn it into check+throw..

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java (hunk 1 above)
Comment (abaranau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java]: fixing

**Pull request merged**